File: backend/src/main/java/dev/coldhands/pair/stairs/backend/usecase/CoreCombinationCalculationService.java
package dev.coldhands.pair.stairs.backend.usecase;

import dev.coldhands.pair.stairs.backend.domain.CombinationCalculationService;
import dev.coldhands.pair.stairs.backend.domain.ScoredCombination;
import dev.coldhands.pair.stairs.backend.infrastructure.mapper.ScoredCombinationMapper;
import dev.coldhands.pair.stairs.backend.infrastructure.persistance.entity.DeveloperEntity;
import dev.coldhands.pair.stairs.backend.infrastructure.persistance.entity.StreamEntity;
import dev.coldhands.pair.stairs.backend.infrastructure.persistance.repository.DeveloperRepository;
import dev.coldhands.pair.stairs.backend.infrastructure.persistance.repository.StreamRepository;
import dev.coldhands.pair.stairs.core.domain.pairstream.PairStream;
import dev.coldhands.pair.stairs.core.usecases.pairstream.PairStreamEntryPoint;

import java.util.List;
import java.util.Map;

import static java.util.stream.Collectors.toMap;

public class CoreCombinationCalculationService implements CombinationCalculationService {

    private final DeveloperRepository developerRepository;
    private final StreamRepository streamRepository;
    private final EntryPointFactory entryPointFactory;

    public CoreCombinationCalculationService(DeveloperRepository developerRepository, StreamRepository streamRepository, EntryPointFactory entryPointFactory) {
        this.developerRepository = developerRepository;
        this.streamRepository = streamRepository;
        this.entryPointFactory = entryPointFactory;
    }

    @Override
    public List<ScoredCombination> calculate(List<Long> developerIds, List<Long> streamIds, int page, int pageSize) {
        // todo validate that ids are real?

        final PairStreamEntryPoint entryPoint = entryPointFactory.create(
                asStrings(developerIds),
                asStrings(streamIds)
        );

        final List<dev.coldhands.pair.stairs.core.domain.ScoredCombination<PairStream>> scoredCombinations = entryPoint.computeScoredCombinations();

        final Map<Long, DeveloperEntity> developerLookup = developerRepository.findAllById(developerIds).stream()
                .collect(toMap(DeveloperEntity::getId, d -> d));
        final Map<Long, StreamEntity> streamLookup = streamRepository.findAllById(streamIds).stream()
                .collect(toMap(StreamEntity::getId, s -> s));

        return scoredCombinations.stream()
                .skip((long) page * pageSize)
                .limit(pageSize)
                .map(sc -> ScoredCombinationMapper.coreToDomain(sc, developerLookup::get, streamLookup::get))
                .toList();
    }

    private static List<String> asStrings(List<Long> ids) {
        return ids.stream().map(Object::toString).toList();
    }
}

File: backend/src/main/kotlin/dev/coldhands/pair/stairs/backend/usecase/CoreCombinationCalculationService.kt
package dev.coldhands.pair.stairs.backend.usecase

import dev.coldhands.pair.stairs.backend.domain.DeveloperId
import dev.coldhands.pair.stairs.backend.domain.Page
import dev.coldhands.pair.stairs.backend.domain.StreamId
import dev.coldhands.pair.stairs.backend.domain.combination.CombinationCalculationService
import dev.coldhands.pair.stairs.backend.domain.combination.PairStream
import dev.coldhands.pair.stairs.backend.domain.combination.ScoredCombination

class CoreCombinationCalculationService(
    private val entryPointFactory: EntryPointFactory
) : CombinationCalculationService {
    override fun calculate(
        developerIds: Collection<DeveloperId>,
        streamIds: Collection<StreamId>,
        page: Int,
        pageSize: Int
    ): Page<ScoredCombination> {
        // todo validate that ids are real?

        val entryPoint = entryPointFactory.create(
            developerIds.map { it.value.toString() },
            streamIds.map { it.value.toString() }
        )

        val scoredCombinations = entryPoint.computeScoredCombinations()

        val pages = scoredCombinations.windowed(size = pageSize, step = pageSize, partialWindows = true)

        if (page < 0 || page >= pages.size) {
            return Page.empty()
        }

        return Page(
            metadata = Page.Metadata(
                nextPageNumber = (page + 1).takeIf { it < pages.size },
            ),
            data = pages
                .drop(page)
                .first()
                .map { sc ->
                    ScoredCombination(
                        score = sc.totalScore,
                        combination = sc.combination.pairs.map { pairStream ->
                            PairStream(
                                developerIds = pairStream.developers().map { DeveloperId(it.toLong()) }.toSet(),
                                streamId = StreamId(pairStream.stream.toLong())
                            )
                        }.toSet()
                    )
                }
        )
    }
}